### PR TITLE
JIT: fix decompose long left shift for overshift cases

### LINKED ISSUE #15566: JIT: GitHub_15291 test failure with MinOpts
Related to #15442 

```
C:\gh\coreclr\bin\tests\Windows_NT.x86.Checked\JIT\Regression\JitBlue\GitHub_15291\GitHub_15291\GitHub_15291.cmd -coreroot C:\gh\coreclr\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root
BEGIN EXECUTION
 "C:\gh\coreclr\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root\corerun.exe" GitHub_15291.exe
r64:FFFFFFFF r64a:00000000
r63:FFFFFFFF r63a:FFFFFFFF
r32:FFFFFFFF r32a:FFFFFFFF
r31:7FFFFFFF r31a:7FFFFFFF
rm1:FFFFFFFF rm1a:FFFFFFFF
Expected: 100
Actual: 0
END EXECUTION - FAILED
FAILED
PASS: 0, FAIL: 1, TOTAL: 1
```

https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x86_checked_windows_nt_minopts/

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Need to reduce the shift amount modulo 64 to match the helper and
`gtFoldExpr` behavior. Since reduced amount must be less than 64
we can remove handling for that case.

Re-enable the test case disabled by #15567.

Closes #15566.

### CODE REVIEW

File: src/jit/decomposelongs.cpp [hunk 1]
```diff
@@ -1112,6 +1112,24 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
         {
             case GT_LSH:
             {
+                // Reduce count modulo 64 to match behavior found in
+                // the LLSH helper and in gtFoldExpr.
+                count &= 0x3F;
```
Comment (mikedn) [inline src/jit/decomposelongs.cpp]: Shouldn't this fix be enabled only on x86? AFAIR arm32 has a different behavior.

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib 

No diffs across FX assemblies. Re-enabled test case diff:
```diff
 ; Assembly listing for method P:G64():int
 ; Emitting BLENDED_CODE for generic X86 CPU
 ; compiler->opts.MinOpts() is true
@@ -18,10 +18,8 @@ G_M3851_IG01:
 G_M3851_IG02:
        C745FC01000000 mov      dword ptr [ebp-04H], 1
        8B45FC       mov      eax, dword ptr [ebp-04H]
-       B81F000000   mov      eax, 31
-       33C0         xor      eax, eax
-       33D2         xor      edx, edx
-       8BC2         mov      eax, edx
+       8B55FC       mov      edx, dword ptr [ebp-04H]
+       C1FA1F       sar      edx, 31
        83E801       sub      eax, 1

 G_M3851_IG03:
@@ -29,11 +27,11 @@ G_M3851_IG03:
        5D           pop      ebp
        C3           ret

-; Total bytes of code 32, prolog size 4 for method P:G64():int
+; Total bytes of code 27, prolog size 4 for method P:G64():int
 ; ============================================================
-r64:FFFFFFFF r64a:00000000
+r64:00000000 r64a:00000000
 r63:FFFFFFFF r63a:FFFFFFFF
 r32:FFFFFFFF r32a:FFFFFFFF
 r31:7FFFFFFF r31a:7FFFFFFF
 rm1:FFFFFFFF rm1a:FFFFFFFF
-ERRORLEVEL=0
+ERRORLEVEL=100
```

File: src/jit/decomposelongs.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/jit/decomposelongs.cpp]: `gtFoldExpr` doesn't do anything different for arm32 / x86, it always reduces. Also the left shift helper for arm uses C++ unsigned left shift semantics, which also reduce.

So I think we're ok with this change on arm32 too, but I'll investigate further.

**Comment by BruceForstall** [comment]
@dotnet-bot test Windows_NT x86 Checked minopts
@dotnet-bot test Windows_NT x86 Checked jitstress1

File: src/jit/decomposelongs.cpp (hunk 1 above)
Comment (mikedn) [reply src/jit/decomposelongs.cpp]: > Also the left shift helper for arm uses C++ unsigned left shift semantics, which also reduce.

The behavior of C++ shift operations when the shift count exceeds the bid width is undefined.

Comment (BruceForstall) [reply src/jit/decomposelongs.cpp]: The arm register LSL instruction uses the low byte of its register count argument as the shift count; slightly different from xarch.

[the remark this answers is not in the document]
Comment (mikedn) [reply src/jit/decomposelongs.cpp]: Hmm, but why does all this applies only to `GT_LSH`? It looks like the x86 `JIT_LLsh` helper was changed to mask the shift count to 6 bits but `JIT_LRsh` and `JIT_LRsz` were not changed. This looks like unnecessary complication, while the IL shift instructions exhibit undefined behavior in this circumstances I doubt that making them behave inconsistently is good.

Comment (mikedn) [reply src/jit/decomposelongs.cpp]: > gtFoldExpr doesn't do anything different for arm32 / x86, it always reduces.

Ah, forgot about that little detail. `ValueNumStore::EvalOpIntegral` never reduces so you get whatever the C++ compiler thinks it's cool.

[the remark this answers is not in the document]
Comment (AndyAyersMS) [reply src/jit/decomposelongs.cpp]: Arguably a similar effort should be applied to the right shifts. Eg `gtFoldExpr` masks those shift counts too.

Not sure if I want to fold those changes in here as getting left shift consistent seems like a reasonable first step.

[the remark this answers is not in the document]
Comment (mikedn) [reply src/jit/decomposelongs.cpp]: I can take care of other cases. You probably have more interesting/useful things to do than fixing things that are ultimately not really broken, just inconsistent/risky :smile:

Comment (AndyAyersMS) [reply src/jit/decomposelongs.cpp]: Seems like we should also update the C++ LLSH helper to mask with 0x3F to be consistent  with non-helper usage and avoid the potential of relying on compiler-defined behavior.

**Comment by AndyAyersMS** [comment]
[armlb failure] is known #15718 

Retrying [minopts](https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x86_checked_windows_nt_minopts_prtest/31/) and [jitstress1](https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x86_checked_windows_nt_jitstress1_prtest/29/) after another known issue #15710 was fixed.

@dotnet-bot test Windows_NT x86 Checked minopts
@dotnet-bot test Windows_NT x86 Checked jitstress1

**Comment by AndyAyersMS** [comment]
For arm32, with minopts, for the GitHub_15291 test case:
```asm
;; before  (returns 0xFFFFFFFF)
; Assembly listing for method P:G64():int
; Emitting BLENDED_CODE for generic ARM CPU
; compiler->opts.MinOpts() is true
; r11 based frame
; partially interruptible
; Final local variable assignments
;
;# V00 OutArgs      [V00    ] (  1,  1   )  lclBlk ( 0) [sp+0x00]   do-not-enreg[]
;  V01 tmp1         [V01,T00] (  3,  6   )     int  ->  [sp+0x04]   do-not-enreg[]
;
; Lcl frame size = 8

G_M3851_IG01:
000000  E92D 4C1C      push    {r2,r3,r4,r10,r11,lr}
000004  F10D 0B10      add     r11, sp, 16

G_M3851_IG02:
000008  2001           movs    r0, 1
00000A  9001           str     r0, [sp+0x04]	// [V01 tmp1]
00000C  9801           ldr     r0, [sp+0x04]	// [V01 tmp1]
00000E  201F           movs    r0, 31
000010  2000           movs    r0, 0
000012  2300           movs    r3, 0
000014  4618           mov     r0, r3
000016  1E40           subs    r0, r0, 1

G_M3851_IG03:
000018  E8BD 8C1C      pop     {r2,r3,r4,r10,r11,pc}

;; after (returns 0x0)
; Assembly listing for method P:G64():int
; Emitting BLENDED_CODE for generic ARM CPU
; compiler->opts.MinOpts() is true
; r11 based frame
; partially interruptible
; Final local variable assignments
;
;# V00 OutArgs      [V00    ] (  1,  1   )  lclBlk ( 0) [sp+0x00]   do-not-enreg[]
;  V01 tmp1         [V01,T00] (  3,  6   )     int  ->  [sp+0x04]   do-not-enreg[]
;
; Lcl frame size = 8

G_M3851_IG01:
000000  E92D 4C1C      push    {r2,r3,r4,r10,r11,lr}
000004  F10D 0B10      add     r11, sp, 16

G_M3851_IG02:
000008  2001           movs    r0, 1
00000A  9001           str     r0, [sp+0x04]	// [V01 tmp1]
00000C  9801           ldr     r0, [sp+0x04]	// [V01 tmp1]
00000E  9B01           ldr     r3, [sp+0x04]	// [V01 tmp1]
000010  17DB           asrs    r3, r3, 31
000012  1E40           subs    r0, r0, 1

G_M3851_IG03:
000014  E8BD 8C1C      pop     {r2,r3,r4,r10,r11,pc}
```
with optimization enabled and the shift amount obscured
```asm
; Assembly listing for method P:Gx(int):int
; Emitting BLENDED_CODE for generic ARM CPU
; optimized code
; r11 based frame
; partially interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T01] (  3,  3   )     int  ->   r2
;# V01 OutArgs      [V01    ] (  1,  1   )  lclBlk ( 0) [sp+0x00]
;  V02 rat0         [V02,T02] (  2,  4   )     int  ->   r0
;  V03 rat1         [V03,T03] (  2,  4   )     int  ->   r1
;  V04 rat2         [V04,T00] (  3,  6   )    long  ->  [sp+0x00]   multireg-ret
;
; Lcl frame size = 8

G_M51697_IG01:
000000  E92D 480C      push    {r2,r3,r11,lr}
000004  F10D 0B08      add     r11, sp, 8
000008  4602           mov     r2, r0

G_M51697_IG02:
00000A  2001           movs    r0, 1
00000C  2100           movs    r1, 0
00000E  F240 0300      movw    r3, LOW RELOC 0x7301dc8
000012  F2C0 0300      movt    r3, HIGH RELOC 0x7301dc8
000016  681B           ldr     r3, [r3]
000018  4798           blx     r3               // CORINFO_HELP_LLSH
00001A  9000           str     r0, [sp]
00001C  9101           str     r1, [sp+0x04]
00001E  9800           ldr     r0, [sp]
000020  9B01           ldr     r3, [sp+0x04]
000022  1E40           subs    r0, r0, 1

G_M51697_IG03:
000024  E8BD 880C      pop     {r2,r3,r11,pc}
```
Unless the LLSH helper masks it will return 0 (as arm LSL interprets shift counts >= 32 as setting result to zero), so the `Gx(64)` result will be `0xFFFFFFFF`.

**Comment by AndyAyersMS** [comment]
@dotnet-bot test Windows_NT x86 Checked minopts
@dotnet-bot test Windows_NT x86 Checked jitstress1
@dotnet-bot test Windows_NT arm Cross Checked minopts Build and Test

**Comment by AndyAyersMS** [comment]
Hmm, looks like I should add GitHub_15291 into the arm/arm64 test list. 

@BruceForstall this is just a manual process, right?

**Comment by BruceForstall** [comment]
Yes, edit tests\arm\Tests.lst and tests\arm64\tests.lst.

(jashook has automation to recreate these automatically, but hasn't run it recently)

File: src/vm/jithelpers.cpp [hunk 2]
```diff
@@ -456,7 +456,7 @@ HCIMPLEND
 HCIMPL2_VV(UINT64, JIT_LLsh, UINT64 num, int shift)
 {
     FCALL_CONTRACT;
-    return num << shift;
+    return num << (shift & 0x3F);
```
Comment (mikedn) [inline src/vm/jithelpers.cpp]: So the general plan is to have everything (except the actual ARM32 shift instructions, of course) mask the shift count, right?

Comment (AndyAyersMS) [reply src/vm/jithelpers.cpp]: Yes. I think that's the best way to ensure that all various compilation modes and architectures give back the same result, and it mirrors what C# does.

**Comment by AndyAyersMS** [comment]
@dotnet-bot test Windows_NT arm Cross Checked minopts Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked minopts Build and Test
@dotnet-bot test Windows_NT x86 Checked minopts

**Comment by AndyAyersMS** [comment]
[arm64](https://ci.dot.net/job/dotnet_coreclr/job/master/job/arm64_cross_checked_windows_nt_flow_prtest/446/) infrastructure failure:
```
ERROR: Cannot delete workspace: remote file operation failed: C:\Users\robox\j\workspace\arm64_cross_c---e6cb65c4 at hudson.remoting.Channel@c2dc869:JNLP4-connect connection from 131.107.19.149/131.107.19.149:61956: java.io.IOException: Unable to delete 'C:\Users\robox\j\workspace\arm64_cross_c---e6cb65c4\bin\Product\Windows_NT.arm64.Checked\x64\PDB'. Tried 3 times (of a maximum of 3) waiting 0.1 sec between attempts.
```

@dotnet-bot test Windows_NT arm64 Cross Checked Innerloop Build and Test

**Comment by AndyAyersMS** [comment]
Looks like the LSH tests now run on arm and pass. So hopefully this settles things for left shifts.

In retrospect it might be simpler to just add the mask in the importer (if there's not one already, perhaps) as that should cover all the cases.

**Pull request merged**